lib/stdlib/test/property_test/queue_prop.erl

```diff
@@ -0,0 +1,670 @@
+		    equal(lists:reverse(lists:delete(Y, lists:reverse(L))), queue:delete_r(Y, Q))
+	    end,
+	    R2 = equal(lists:reverse(lists:delete(X, lists:reverse(L))), queue:delete_r(X, Q)),
+
+	    R1 andalso R2
+	end
+    ).
+
+prop_delete_with() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    Q1 = queue:delete_with(fun is_atom/1, Q),
+	    L1 = case lists:search(fun is_atom/1, L) of
+		false ->
+		    L;
+		{value, V} ->
+		    lists:delete(V, L)
+	    end,
+	    equal(L1, Q1)
+	end
+    ).
+
+prop_delete_with_r() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    Q1 = queue:delete_with_r(fun is_atom/1, Q),
+	    L1 = lists:reverse(L),
+	    L2 = case lists:search(fun is_atom/1, L1) of
+		false ->
+		    L;
+		{value, V} ->
+		    lists:reverse(lists:delete(V, L1))
+	    end,
+	    equal(L2, Q1)
+	end
+    ).
+
+prop_drop() ->
+    common_drop_tail(drop).
+
+prop_drop_r() ->
+    common_drop_r_init_liat(drop_r).
+
+prop_filter() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    Q1 = queue:filter(
+		fun
+		    (I) when is_atom(I) -> true;
+		    (I) when is_integer(I) -> [I * 2];
+		    (I) when is_float(I) -> [{I, I}];
+		    (I) when is_tuple(I) -> [I, I];
+		    (_) -> false
+		end,
+		Q
+	    ),
+	    L1 = lists:foldr(
+		fun
+		    (I, Acc) when is_atom(I) -> [I|Acc];
+		    (I, Acc) when is_integer(I) -> [I * 2|Acc];
+		    (I, Acc) when is_float(I) -> [{I, I}|Acc];
+		    (I, Acc) when is_tuple(I) -> [I, I|Acc];
+		    (_, Acc) -> Acc
+		end,
+		[],
+		L
+	    ),
+	    equal(L1, Q1)
+	end
+    ).
+
+prop_filtermap() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    F = fun
+		(I) when is_atom(I) ->
+		    true;
+		(I) when is_integer(I) ->
+		    {true, {I, I}};
+		(_) ->
+		    false
+	    end,
+	    Q1 = queue:filtermap(F, Q),
+	    L1 = lists:filtermap(F, L),
+	    equal(L1, Q1)
+	end
+    ).
+
+prop_fold() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    % order-independent fold
+	    F1 = fun
+		(I, Acc) when is_number(I) -> Acc + 2 * I;
+		(_, Acc) -> Acc
+	    end,
+	    RQ1 = queue:fold(F1, 0, Q),
+	    RL1 = lists:foldl(F1, 0, L),
+
+	    % order-dependent fold
+	    F2 = fun
+		(I, Acc) -> [{I, I}|Acc]
+	    end,
+	    RQ2 = queue:fold(F2, [], Q),
+	    RL2 = lists:foldl(F2, [], L),
+
+	    RQ1 =:= RL1 andalso
+	    RQ2 =:= RL2
+	end
+    ).
+
+prop_get() ->
+    common_get_head(get).
+
+prop_get_r() ->
+    common_get_r_last_daeh(get_r).
+
+prop_head() ->
+    common_get_head(head).
+
+prop_in() ->
+    ?FORALL(
+	L,
+	list(),
+	begin
+	    Q = lists:foldl(
+		fun (I, Acc) ->
+		    queue:in(I, Acc)
+		end,
+		queue:new(),
+		L
+	    ),
+	    equal(L, Q)
+	end
+    ).
+
+prop_in_r() ->
+    common_in_r_cons(in_r).
+
+prop_init() ->
+    common_drop_r_init_liat(init).
+
+prop_is_empty() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    (length(L) =:= 0) =:= queue:is_empty(Q)
+	end
+    ).
+
+prop_join() ->
+    ?FORALL(
+	{{L1, Q1}, {L2, Q2}},
+	{list_queue(), list_queue()},
```

Comment:
Hm, ok, I'll see to that :blush: But I guess I'll have the internals of the generator create two lists (instead of splitting one), create the queue from them as you suggest, and return it together with the concatenation of the two "source" lists. Would that be ok?

Reply:
Yes, actual implementation does not really matter.  My main point was to generate _different_ internal representations to make the testing more complete.